### DISCUSSION TITLE
Request for Custom Block SDK/API Support

### DISCUSSION THREAD

**Discussion by gitacoco** [GitbookIO/community#778] [Feature Requests]
Hello,

I'm requesting Gitbook to consider implementing an SDK or API that would allow developers to create custom blocks.

As a design system lead in our department, we have been exploring various tools for design system documentation. Although we are currently using Zeroheight and Supernova, neither tool matches the reliability and modern aspects of Gitbook. The only reason we have not chosen Gitbook is that they offer many design-specific blocks that Gitbook does not. 

Gitbook's current block system, while solid, doesn't fully address domain-specific needs like interactive color ramps with copy-to-clipboard functionality. To address this gap, I would like to propose that Gitbook supports an SDK or API for creating custom blocks. 

Why This Matters:
- Platform Scalability: An SDK/API for custom blocks would transform Gitbook into a documentation operating system, capable of serving diverse documentation needs across different domains.
- Community Empowerment: Rather than Gitbook developing every specialized block internally, a marketplace approach would leverage community creativity and expertise.
- Competitive Advantage: While design-specific tools exist, they often struggle with reliability and core documentation features. Gitbook could uniquely position itself by combining its robust platform with customizable domain-specific functionality. Design is another field that has strong documentation needs than engineering. Additionally, this feature would not only benefit design field but would also open up possibilities for other specialized documentation needs across different industries and use cases. This is a big revenue growth potential.

Impact of not being able to have this feature:
There have been multiple occasions where I’ve considered migrating our design system documentation to Gitbook due to its impressive features. However, without the ability to create or access these domain-specific blocks, it's challenging to justify the switch, despite the platform's strengths.

**Comment by SamyPesse** [comment]
Hey @gitacoco, good news is that we have an API / Integration platform that support building custom blocks: https://developer.gitbook.com/integrations/integrations

For example, you can find here the Linear block provided by our Linear integrartion: https://github.com/GitbookIO/integrations/blob/main/integrations/linear/src/index.tsx#L82

Currently the integration platform is not perfectly documented and a bit rough on the edge to work with, but we are constantly improving it and its documentation.

The way it works:

1. You can create an integration, public (for all GitBook users to discover) or private (only your organization can install it)
2. The integration can provide a set of custom blocks (ex: [for Mermaid](https://github.com/GitbookIO/integrations/blob/main/integrations/mermaid/gitbook-manifest.yaml#L25-L31))
3. Each custom block is associated to a ContentKit component. ContentKit components are rendered on the server using our own UI syntax, they can't directly render things in the DOM, but can render a [wide set of elements](https://developer.gitbook.com/integrations/contentkit/reference/reference), respond to user actions, render iframe, etc.
4. Then after installing the integration, you can directly insert the block from the editor or reference it in Markdown

If you are interested in building with it, let us know and we can reach out to help you with @addisonschultz .
